noir_stdlib/src/array.nr

```diff
@@ -1,16 +1,5 @@
 #[builtin(array_len)]
 fn len<T>(_input : [T]) -> comptime Field {}
 
-// insertion sort - n.b. it is a quadratic sort
-fn sort<T>(mut a: [T]) -> [T] { 
-    for i in 1..len(a) {
-        for j in 0..i {
-            if(a[i] < a[j]) {
-                let c = a[j];
-                a[j] = a[i];
-                a[i]= c;
-            }
-        };
-    };
-    a
-}
+#[builtin(arraysort)]
+ fn sort<T>(_a: [T]) -> [T] {}
```

Comment:
I still hesitate to keep this function, and especially hesitate to make a builtin out of it. The original function was temporary before we had higher order functions since it assumes an ordering (via `<`) that doesn't work for all types or isn't always desired. We currently do not implement them but arrays of structs, or nested arrays should be a goal that we shouldn't need a specialized sorting function for.

A more ideal implementation would work for all element types and allow users to sort in descending order or any other they may choose:
```rs
fn sort<T>(mut a: [T], ordering: fn(T, T) -> bool) -> [T] { 
    for i in 1..len(a) {
        for j in 0..i {
            if ordering(a[i], a[j]) {
                let c = a[j];
                a[j] = a[i];
                a[i]= c;
            }
        }
    }
    a
}
```

Reply:
I don't think that builtin/not builtin is relevant here.
I agree that a sort with ordering will be desirable once we support nested arrays of structs. We will be able to support some customisation, that is the user can specify an order by on the structure fields: order by this field, then this one, etc...
Full custom ordering will not be possible, and in that case he will have to write the sorting method by himself, which is easy thanks to high order functions.
I am fine with these limitations.

Reply:
In fact custom ordering is still possible in theory but would require some additional machinery.

Reply:
I resolve the conversation as we agreed to solve this with a new generic integer type that would support comparisons. But it will be done in another PR.